common/core/test/_promise_utils_test.js

```diff
@@ -0,0 +1,565 @@
+
+describe('PromiseUtils', () => {
+    describe('#callbackToPromise', () => {
+        it('executes user callback when passed', function (done) {
+            const functionWithCallback = (callback) => {
+                callback(undefined, 42);
+            };
+
+            const callback = (error, result) => {
+                assert.equal(result, 42);
+                assert.isUndefined(error);
+                done();
+            };
+
+            const promise = callbackToPromise(functionWithCallback, callback);
+            assert.isUndefined(promise);
+        });
+
+        it('throws when user callback is not a Function', function(done) {
+            const functionWithCallback = (callback) => {
+                callback();
+            };
+            const userCallback = 42;
+
+            assert.throws(function() {
+                callbackToPromise(functionWithCallback, userCallback);
+                done("It should never reach this code")
+            }, TypeError);
+
+            done();
+        });
+
+        it('throws when an Error is thrown', function() {
+            const functionWithCallback = (_) => {
+                throw new Error();
+            };
+
+            callbackToPromise(functionWithCallback)
+                .then(res => assert.fail(res), err => assert.isDefined(err));
+        });
+
+        it('returns undefined result for empty callback', function (done) {
+            const functionWithEmptyPromise = (callback) => {
+                callback();
+            };
+
+            callbackToPromise(functionWithEmptyPromise).then(result => {
+                assert.isUndefined(result);
+                done();
+            }).catch(error => {
+                done(error);
+            });
+        });
+
+        it('returns undefined result for callback with undefined result', function (done) {
+            const functionWithEmptyPromise = (callback) => {
+                callback(undefined, undefined);
+            };
+
+            callbackToPromise(functionWithEmptyPromise).then(result => {
+                assert.isUndefined(result);
+                done();
+            }).catch(error => {
+                done(error);
+            });
+        });
+
+        it('returns simple value properly when callback invoked', function (done) {
+            const returnValue = 'sample value';
+            const functionWithSimpleResult = (callback) => {
+                callback(undefined, returnValue);
+            };
+
+            callbackToPromise(functionWithSimpleResult).then(result => {
+                assert.equal(result, returnValue);
+                done();
+            }).catch(error => {
+                done(error);
+            });
+        });
+
+        it('returns complex object properly when callback invoked', function (done) {
+            const returnValue = {
+                key: 'value',
+                id: 42
+            };
+            const functionWithSimpleResult = (callback) => {
```

Comment:
>Simple [](start = 30, length = 6)

nit: complex.

Reply:
I guess I should have mentioned the other Complex ones in this file...

---
In reply to: [222498676](https://github.com/Azure/azure-iot-sdk-node/pull/372#discussion_r222498676) [](ancestors = 222498676)